xbmc/interfaces/json-rpc/PlayerOperations.cpp

```diff
@@ -489,6 +490,153 @@ JSONRPC_STATUS CPlayerOperations::Zoom(const std::string &method, ITransportLaye
   }
 }
 
+JSONRPC_STATUS CPlayerOperations::SetViewMode(const std::string &method, ITransportLayer *transport, IClient *client, const CVariant &parameterObject, CVariant &result)
+{
+  ViewMode mode;
+  // init with current values from settings
+  CVideoSettings vs = g_application.GetAppPlayer().GetVideoSettings();
+
+  CVariant viewMode = parameterObject["viewmode"];
+  if (viewMode.isString())
+  {
+    std::string modestr = viewMode.asString();
+    if (modestr == "normal")
+      mode = ViewModeNormal;
+    else if (modestr == "zoom")
+      mode = ViewModeZoom;
+    else if (modestr == "stretch4x3")
+      mode = ViewModeStretch4x3;
+    else if (modestr == "widezoom")
+      mode = ViewModeWideZoom;
+    else if (modestr == "stretch16x9")
+      mode = ViewModeStretch16x9;
+    else if (modestr == "original")
+      mode = ViewModeOriginal;
+    else if (modestr == "stretch16x9nonlin")
+      mode = ViewModeStretch16x9Nonlin;
+    else if (modestr == "zoom120width")
+      mode = ViewModeZoom120Width;
+    else if (modestr == "zoom110width")
+      mode = ViewModeZoom110Width;
+    else
+      return InvalidParams;
+  }
+  else if (viewMode.isObject())
+  {
+    mode = ViewModeCustom;
+    CVariant zoom = viewMode["zoom"];
+    CVariant pixelRatio = viewMode["pixelratio"];
+    CVariant verticalShift = viewMode["verticalshift"];
+    CVariant stretch = viewMode["nonlinearstretch"];
+    if (zoom.isNull() && pixelRatio.isNull() && verticalShift.isNull() && stretch.isNull())
+      return InvalidParams;
+
+    if (zoom.isDouble())
+      vs.m_CustomZoomAmount = zoom.asDouble();
+    else if (zoom.isString())
+    {
+      std::string strZoom = zoom.asString();
+      if (strZoom == "increase")
+      {
+        vs.m_CustomZoomAmount += 0.01f;
+        if (vs.m_CustomZoomAmount > 2.f)
+          vs.m_CustomZoomAmount = 2.f;
+      }
+      else if (strZoom == "decrease")
+      {
+        vs.m_CustomZoomAmount -= 0.01f;
+        if (vs.m_CustomZoomAmount < 0.5f)
+          vs.m_CustomZoomAmount = 0.5f;
+      }
+      else
+        return InvalidParams;
+    }
+
+    if (pixelRatio.isDouble())
+      vs.m_CustomPixelRatio = pixelRatio.asDouble();
+    else if (pixelRatio.isString())
+    {
+      std::string strPixelRatio = pixelRatio.asString();
+      if (strPixelRatio == "increase")
+      {
+        vs.m_CustomPixelRatio += 0.01f;
+        if (vs.m_CustomPixelRatio > 2.f)
+          vs.m_CustomPixelRatio = 2.f;
+      }
+      else if (strPixelRatio == "decrease")
+      {
+        vs.m_CustomPixelRatio -= 0.01f;
+        if (vs.m_CustomPixelRatio < 0.5f)
+          vs.m_CustomPixelRatio = 0.5f;
+      }
+      else
+        return InvalidParams;
+    }
+
+    if (verticalShift.isDouble())
+      vs.m_CustomVerticalShift = verticalShift.asDouble();
+    else if (verticalShift.isString())
+    {
+      std::string strVerticalShift = verticalShift.asString();
+      if (strVerticalShift == "increase")
```

Comment:
Increase and decrease looks mixed up here?

Reply:
Don‘t worry about it - I will refactor the c++ part to show you how to prevent that bunch of code duplication.

Reply:
Maybe. My thinking was vertical shift "increase" would do same as ACTION_VSHIFT_UP, but that does decrease vertical shift so possibly confusing. Do people know vertical shift increases downwards?

It is better in JSON to have all parameters use "increase"/"decrease" than have "in"/"out" for zoom, "up"/"down" for vertical shift and "increase"/"decrease" for pixel ratio used in one method call.